### PR TITLE
OPENINSAI: SWI-4549 - Prepare saiswitch.py file for upstreaming

### PR DESCRIPTION
**Pull request by pkochmanska**

### CODE REVIEW

File: ptf/saiswitch.py [hunk 1]
```diff
@@ -0,0 +1,1642 @@
...
+from ptf.mask import Mask
+from ptf.packet import *
+from ptf.testutils import *
+from ptf.thriftutils import *
+
+from sai_base_test import *
+
+THIS_DIR = os.path.dirname(os.path.abspath(__file__))
+sys.path.append(os.path.join(THIS_DIR, '..'))
+from common.lpm import *
```
Comment (aljer) [inline ptf/saiswitch.py]: I'm not sure if we upstream any modules from common directory. I see two solutions: 
1) to add this file 
2) [imo better] to add LpmDict class from lpm.py to our sai_utils.py module

File: ptf/saiswitch.py [hunk 2]
```diff
@@ -0,0 +1,1642 @@
...
+
+        Logic to check if one extra entry can be programmed is not
+        correct. The available LPM value returned by SAI API will return
+        85% of total LPM table size. Its a recomm value i.e. not to go
+        above 85% of total LPM routes as there can be Hash collisions at
+        some point. But that doesn't mean the return max value is the
+        fixed value, you can always go above this number if the IP prefix
+        combination is not resulting in hash collision. Similarly there
+        can be scenarios where hash collision can happen even before
+        reaching max LPM routes [i.e. 85% of LPM table size].
```
Comment (aljer) [inline ptf/saiswitch.py]: I think this was our internal issue

File: ptf/saiswitch.py [hunk 3]
```diff
@@ -0,0 +1,1642 @@
...
```
Comment (aljer) [inline ptf/saiswitch.py]: I think this was our internal issue

File: ptf/saiswitch.py [hunk 4]
```diff
@@ -0,0 +1,1642 @@
...
+
+    def availableIPv4NexthopEntryTest(self):
+        '''
+        Verifies creation of maximum number of IPv4 nexthop entries.
+
+        commenting out the logic to program one extra nexthop
+        cant try to program last nexthop as it will error out because
+        programming the last nexthop i.e. table-max id [i.e. 65536 or
+        32768] as key is not possible. nexthop-id key starts from 1 &
+        not zero, so total available will be table max-entries - 1.
```
Comment (aljer) [inline ptf/saiswitch.py]: I think this was our internal issue

File: ptf/saiswitch.py [hunk 5]
```diff
@@ -0,0 +1,1642 @@
...
+
+    def availableIPv6NexthopEntryTest(self):
+        '''
+        Verifies creation of maximum number of IPv6 nexthop entries.
+
+        commenting out the logic to program one extra nexthop
+        cant try to program last nexthop as it will error out because
+        programming the last nexthop i.e. table-max id [i.e. 65536 or
+        32768] as key is not possible. nexthop-id key starts from 1 &
+        not zero, so total available will be table max-entries - 1.
```
Comment (aljer) [inline ptf/saiswitch.py]: I think this was our internal issue

File: ptf/saiswitch.py [hunk 6]
```diff
@@ -0,0 +1,1642 @@
...
+    def availableIPv4NeighborEntryTest(self):
+        '''
+        Verifies creation of maximum number of IPv4 neighbor entries.
+
+        commenting out the logic to program one extra neighbor
+        cant try to program last neighbor as it will error out because
+        it will internally try to prog last nexthop i.e. table-max id
+        [i.e. 65536 or 32768] as key which is not possible. nexthop-id
+        key starts from 1 & not zero, so total available will be table
+        max-entries - 1
```
Comment (aljer) [inline ptf/saiswitch.py]: I think this was our internal issue

File: ptf/saiswitch.py [hunk 7]
```diff
@@ -0,0 +1,1642 @@
...
+    def availableIPv6NeighborEntryTest(self):
+        '''
+        Verifies creation of maximum number of IPv6 neighbor entries.
+
+        commenting out the logic to program one extra neighbor
+        cant try to program last neighbor as it will error out because
+        it will internally try to prog last nexthop i.e. table-max id
+        [i.e. 65536 or 32768] as key which is not possible. nexthop-id
+        key starts from 1 & not zero, so total available will be table
+        max-entries - 1
```
Comment (aljer) [inline ptf/saiswitch.py]: I think this was our internal issue

File: ptf/saiswitch.py [hunk 8]
```diff
@@ -0,0 +1,1642 @@
...
+                        available_next_hop_group_member_entry=True)
+                    self.assertEqual(
+                        attr["available_next_hop_group_member_entry"],
+                        max_member_entry - member_number)
+
+            # Max number of ecmp groups is max-entries - 1, entry 0 is reserved
+            # Since we have one less ecmp entry, we cant use all available ecmp
+            # members as we can have only 64 members per group in current
+            # profiles. We will always see 64-1 [default] ecmp members left out
+            # so just check if we were able to program max ecmp group
```
Comment (aljer) [inline ptf/saiswitch.py]: Barefoot-specific issue.?

File: ptf/saiswitch.py [hunk 9]
```diff
@@ -0,0 +1,1642 @@
...
+@group('hw')
+class SwitchHwAttrTest(SaiHelper):
+    '''
+    Switch hardware-related attributes tests
+    '''
+
+    def runTest(self):
+        self.refreshIntervalTest()
+
+    def refreshIntervalTest(self):
```
Comment (aljer) [inline ptf/saiswitch.py]: We do not distinguish between hw and model tests. This may be moved to SwitchAttrTest class.

File: ptf/saiswitch.py [hunk 10]
```diff
@@ -0,0 +1,1642 @@
...
+
+@group('nat')
+class SwitchNatAttrTest(SaiHelper):
+    '''
+    Switch NAT-related attributes tests
+    '''
+
+    def runTest(self):
+        self.availableSnatEntryTest()
+        self.availableDnatEntryTest()
```
Comment (aljer) [inline ptf/saiswitch.py]: Similarly like for hw test - these may be moved to SwitchAttrTest

File: ptf/saiswitch.py [hunk 11]
```diff
@@ -0,0 +1,1642 @@
...
+            finally:
+                if not stat:
+                    sai_thrift_remove_nat_entry(self.client, dnat)
+
+        finally:
+            for dnat in dnat_list:
+                sai_thrift_remove_nat_entry(self.client, dnat)
+
+
+@group('tunnel')
```
Comment (aljer) [inline ptf/saiswitch.py]: @group decorator is not needed

File: ptf/saiswitch.py (hunk 1 above)
Comment (pkochmanska) [reply ptf/saiswitch.py]: LpmDict to be moved in separate PR

Comment (aljer) [reply ptf/saiswitch.py]: IMO we can still add it to upstreamed sai_utils since it is not merged yet. Let's assume here we have this class in sai_utils.

**Comment by ravi861** [comment]
SwitchDefaultVlanTest which I recently added. I don't see this here. LGTM otherwise.

**Review by ravi861 — APPROVED** [review]
LGTM. Good to go.

File: ptf/saiswitch.py [hunk 12]
```diff
@@ -0,0 +1,1776 @@
...
+        str: unique IP address
+    '''
+    if not ipv6:
+        ip_range = [addr for addr in
+                    ["1.0.0.1", "254.255.255.254"]]
+    else:
+        ip_range = [addr for addr in
+                    ["2001::0", "2001:0db8::ffff:ffff:ffff"]]
+
+    ip_interval = LpmDict.IpInterval(
```
Comment (aljer) [inline ptf/saiswitch.py]: Please add "from lpm import LpmDict" import statement since lpm module was added to tests infrastructure in public PR.

**Review by aljer — APPROVED** [review]
LGTM, please upstream also lpm.py module together with the test file

**Pull request closed without merging**